Automated test plan

edipascale opened the pull request.

Frostman commented:
@edipascale could you please make a separate PR with the things refactoring in the testing.go so we can get it merged quickly and you'll not have to rebase

Frostman commented on .github/workflows/ci.yaml:
```diff
@@ -424,10 +426,28 @@ jobs:
+        run: |
+          source "./lab-ci/envs/$KUBE_NODE/source.sh"
+          bin/hhfab init -v --dev --include-onie=${{ matrix.includeonie }} -w "./lab-ci/envs/$KUBE_NODE/wiring.yaml"
+          bin/hhfab vlab up -v --ready switch-reinstall --ready wait --ready release-test --ready exit --mode=${{ matrix.buildmode }}
```
We should use `--ready inspect` instead of `--ready wait` and I think it'll be great to keep the standard test for uniformity of results (`--ready setup-vpcs --ready test-connectivity`) before running the release tests

Frostman commented on pkg/hhfab/release.go:
```diff
@@ -0,0 +1,1499 @@
+// options in the test context
+func (testCtx *VPCPeeringTestCtx) setupTest(ctx context.Context) error {
+	if err := hhfctl.VPCWipeWithClient(ctx, testCtx.kube); err != nil {
+		return errors.Wrap(err, "WipeBetweenTests")
```
When wrapping error please provide context in a human readable message, assuming migration to a new errors smth like:

`return fmt.Errorf("wiping vpcs between tests: %w", err)`

Frostman commented on pkg/hhfab/release.go:
```diff
@@ -0,0 +1,1499 @@
+		return errors.Wrap(err, "WipeBetweenTests")
+	}
+	if err := DoVLABSetupVPCs(ctx, testCtx.workDir, testCtx.cacheDir, testCtx.opts); err != nil {
+		return errors.Wrap(err, "DoVLABSetupVPCs")
```
Same here, I'll not explicitly mark other occurencies

Frostman commented on pkg/hhfab/release.go:
```diff
@@ -0,0 +1,1499 @@
+func populateFullMeshVpcPeerings(ctx context.Context, kube client.Client, vpcPeerings map[string]*vpcapi.VPCPeeringSpec) error {
+	vpcs := &vpcapi.VPCList{}
+	if err := kube.List(ctx, vpcs); err != nil {
+		return errors.Wrapf(err, "kube.List")
```
For such cases as well, we want to get little bit more context, e.g. "listing vpcs: %w"

Frostman commented on pkg/hhfab/release.go:
```diff
@@ -0,0 +1,1499 @@
+		slog.Debug("Output of ping", "output", string(out))
+
+		return errors.New("ping failed, expected success")
+	} else if !expectSuccess && pingOK {
```
What if we expect ping to fail (expectSuccess is false), but only some of packets are lost (pingOK MAY be false too? In theory ping should return non-zero code which will cause cmd to fail with error, let's confirm it) - in that case we'll just return nil like there is no issues.

Frostman commented on pkg/hhfab/release.go:
```diff
@@ -0,0 +1,1499 @@
+	}
+	defer func() {
+		if err := changeAgentStatus(testCtx.hhfabBin, testCtx.workDir, deviceName, true); err != nil {
+			slog.Error("Failed to enable agent", "error", err)
```
As a TODO, no need to fix now: It feels like we may want to retry it as if we aren't able to bring agent back future tests will fail and/or we'll keep env in a broken state

Frostman commented on pkg/hhfab/release.go:
```diff
@@ -0,0 +1,1499 @@
+	}
+	defer func() {
+		if err := changeSwitchPortStatus(testCtx.hhfabBin, testCtx.workDir, deviceName, nosPortName, true); err != nil {
+			slog.Error("Failed to set port up", "error", err)
```
Not so bad in this case as agent should enforce it eventually

Frostman commented on pkg/hhfab/release.go:
```diff
@@ -0,0 +1,1499 @@
+				if err := changeSwitchPortStatus(testCtx.hhfabBin, testCtx.workDir, spine.Name, nosPortName, false); err != nil {
+					return false, errors.Wrap(err, "failed to set port down")
+				}
+				// XXX: do we need to set ports back up? restarting the agent should eventually take care of that
```
Should it be converted to a TODO or removed?

Frostman commented on pkg/hhfab/release.go:
```diff
@@ -0,0 +1,1499 @@
+	_, err = CreateOrUpdateVpc(ctx, testCtx.kube, vpc2)
+	if err != nil {
+		return false, errors.Wrap(err, "CreateOrUpdateVpc")
+	}
```
Should we test connectivity again afer removing restrictions?

Frostman commented on pkg/hhfab/release.go:
```diff
@@ -0,0 +1,1499 @@
+	_, err = CreateOrUpdateVpc(ctx, testCtx.kube, vpc)
+	if err != nil {
+		return false, errors.Wrap(err, "CreateOrUpdateVpc")
+	}
```
Same here - would be good to test connectivity IMO

Frostman commented on pkg/hhfab/release.go:
```diff
@@ -0,0 +1,1499 @@
+		}
+		slog.Debug("Creating VPCAttachment", "attachment", newVpcAtt.Name)
+		if err := testCtx.kube.Create(ctx, newVpcAtt); err != nil {
+			slog.Error("Error creating VPCAttachment", "error", err)
```
As a TODO, no need to fix now: Failing this defer would keep fabric in the "broken" state

Frostman commented on pkg/hhfab/release.go:
```diff
@@ -0,0 +1,1499 @@
+	}
+	defer func() {
+		slog.Debug("Removing address and default route from en2ps1 on the server", "server", server)
+		_ = execNodeCmd(testCtx.hhfabBin, testCtx.workDir, server, "sudo ip addr del 172.30.50.1/24 dev enp2s1")
```
ignoring the errors from cmd execs will completely hide from is we've been unable to conf ips or cleanup

Frostman commented on pkg/hhfab/release.go:
```diff
@@ -0,0 +1,1499 @@
+		if err := execNodeCmd(testCtx.hhfabBin, testCtx.workDir, "server-1", "/opt/bin/hhnet cleanup"); err != nil {
+			slog.Error("Error cleaning up server-1", "error", err)
+		}
+		// FIXME: ideally this would be derived rather than hardcoded (extract the code from testing.go)
```
Convert to TODO? I think it's okay for this PR

Frostman commented on pkg/hhfab/release.go:
```diff
@@ -0,0 +1,1499 @@
+	if err := execNodeCmd(testCtx.hhfabBin, testCtx.workDir, "server-1", "/opt/bin/hhnet cleanup"); err != nil {
+		return false, errors.Wrap(err, "execNodeCmd")
+	}
+	// FIXME: ideally this would be derived rather than hardcoded (extract the code from testing.go)
```
Convert to TODO? I think it's okay for this PR

Frostman commented on pkg/hhfab/release.go:
```diff
@@ -0,0 +1,1499 @@
+}
+
+func RunReleaseTestSuites(ctx context.Context, workDir, cacheDir string, rtOtps ReleaseTestOpts) error {
+	// FIXME: make this configurable
```
Convert to TODO? I think it's okay for this PR

Frostman commented on pkg/hhfab/release.go:
```diff
@@ -0,0 +1,1499 @@
+
+func printSuiteResults(ts *JUnitTestSuite) {
+	var numFailed, numSkipped, numPassed int
+	slog.Info(fmt.Sprintf("Results for %s:", ts.Name))
```
slog is supposed to be used with args smth like slog.Info("Results", "testSuite", ts.Name)

Frostman commented on pkg/hhfab/release.go:
```diff
@@ -0,0 +1,1499 @@
+	slog.Info(fmt.Sprintf("Results for %s:", ts.Name))
+	for _, test := range ts.TestCases {
+		if test.Skipped != nil {
+			slog.Info(fmt.Sprintf("SKIP %s", test.Name), "skip-message", test.Skipped.Message)
```
Same, and it'll be more clear (and parseable) - `slog.Info("SKIP", "test", test.Name, "reason", test.Skipped.Message)

Frostman commented on pkg/hhfab/release.go:
```diff
@@ -0,0 +1,1499 @@
+			slog.Warn(fmt.Sprintf("FAIL %s", test.Name), "error", strings.Split(test.Failure.Message, "\n")[0])
+			numFailed++
+		} else {
+			slog.Info(fmt.Sprintf("PASS %s", test.Name))
```
Here and later for `slog.Info(fmt.Sprintf`

Frostman reviewed (changes requested):
Great job! I wasn't digging much into each test case, mainly looking for potential issues, unclear behavior, and general Go code issues.

Few notes for fixing in future PRs:
- I have a few questions about making sure that tests that are suitable for the VLAB will run
- failed cleanups in defer may leave fabric in a semi-broken state

edipascale commented on pkg/hhfab/release.go:
```diff
@@ -0,0 +1,1499 @@
+		Link: wiringapi.ConnStaticExternalLink{
+			Switch: wiringapi.ConnStaticExternalLinkSwitch{
+				BasePortName: wiringapi.NewBasePortName(switchPortName),
+				IP:           "172.30.50.5/24",
```
changed to 172.31.255.X

edipascale commented:
@Frostman I think I've addressed all of your comments. A couple of standing issues:
- a run now takes ~2h5m in env-1. I tried cutting corners but short of disabling iperfs (which we agreed we cannot afford to do) the only other option I can think of is to skip entire tests. There is an extended flag which is set to false by default, but currently it only disables curls and some extra steps in a few tests. Infra is already there to skip entire tests based on that flag, if desired.
- vlab release-tests in the ci have been enabled using the same label trick, but based on many runs on my local machine I had to disable several tests, i.e. all failover ones, the static external and all of those with "restrictions". Then there are those that won't run because of the lack of externals, and the final result is that there's only a couple of tests left.
- even on real envs, tests seem to be quite brittle. I've had a few runs with no issues, but many more with one/two random failures, typically on a ping check, pretty much on any of the tests with no discernible pattern. I've also seen one failure with a slightly below target iperf speed (8490 vs 8500). Unless I missed something, we are already waiting for switches to be ready with an "applied for" timer of at least 30 seconds in all instances where I've observed failures, but it doesn't seem to be enough. I could try to increase the timer further, but that would make the runs even longer, and regardless I'm not sure where we should set the acceptance bar or how long is reasonable to wait for

Feel free to have a look and/or let me know if you want to discuss this over a call!

edipascale commented:
one additional data point: after bumping the apppliedFor timer to 1m everywhere, I still got a failure in env-1:
```
10:37:54 ERR FAIL test="Full Mesh All Externals" error="testing connectivity: testing connectivity: testing connectivity from \"server-1\" to \"server-3\": checking ping from server-1 to server-3: should be reachable but ping failed with output: PING 10.10.3.2 (10.10.3.2) 56(84) bytes of data."
[...]
10:37:54 INF All tests completed duration=2h12m32.561923603s
10:37:54 ERR release-test: some tests failed: singleVpc=0, multiVpc=0, basic=1
```
the length of the tests did not increase by too much, I'll try 2 minutes next

edipascale commented:
> the length of the tests did not increase by too much, I'll try 2 minutes next

First run with a 2 minutes timer succeeded, but took considerably longer:
```
16:16:15 INF All tests completed duration=3h0m32.49154262s
```

I'm going to run a few more to see if it's consistent

edipascale commented:
> First run with a 2 minutes timer succeeded, but took considerably longer:
> 
> ```
> 16:16:15 INF All tests completed duration=3h0m32.49154262s
> ```
> 
> I'm going to run a few more to see if it's consistent

it isn't :(
```
19:25:04 INF All tests completed duration=3h1m39.01107962s
19:25:04 ERR release-test: some tests failed: singleVpc=1, multiVpc=0, basic=1
```
